File: desktop-client/pystudio/git_service.py
from __future__ import annotations

import shutil
import subprocess
from dataclasses import dataclass
from pathlib import Path


@dataclass
class GitResult:
    ok: bool
    message: str


class GitService:
    """Conexión Git local del IDE.

    No modifica el backend. El repositorio Git se crea dentro del proyecto local
    del estudiante para guardar historial de cambios del código.
    """

    def is_git_available(self) -> bool:
        return shutil.which("git") is not None

    def init_repo(self, project_root: Path) -> GitResult:
        if not self.is_git_available():
            return GitResult(False, "Git no está instalado o no está en PATH.")
        if (project_root / ".git").exists():
            return GitResult(True, "El repositorio Git local ya existe.")
        result = self._run(project_root, ["git", "init"])
        if not result.ok:
            return result
        self._run(project_root, ["git", "config", "user.email", "pystudio.local@desktop"])
        self._run(project_root, ["git", "config", "user.name", "PyStudio Desktop"])
        return GitResult(True, "Repositorio Git local inicializado.")

    def commit_all(self, project_root: Path, message: str) -> GitResult:
        init = self.init_repo(project_root)
        if not init.ok:
            return init
        self._run(project_root, ["git", "add", "."])
        status = self._run(project_root, ["git", "status", "--porcelain"])
        if not status.message.strip():
            return GitResult(True, "No hay cambios nuevos para commitear.")
        result = self._run(project_root, ["git", "commit", "-m", message])
        if result.ok:
            return GitResult(True, f"Commit creado: {message}")
        return result

    def history(self, project_root: Path) -> str:
        if not (project_root / ".git").exists():
            return "Este proyecto todavía no tiene repositorio Git local."
        result = self._run(project_root, ["git", "log", "--oneline", "--decorate", "-10"])
        return result.message if result.message.strip() else "Aún no hay commits."

    def _run(self, cwd: Path, command: list[str]) -> GitResult:
        try:
            completed = subprocess.run(command, cwd=str(cwd), capture_output=True, text=True)
            output = (completed.stdout + completed.stderr).strip()
            return GitResult(completed.returncode == 0, output)
        except Exception as exc:
            return GitResult(False, f"Error ejecutando Git: {exc}")

File: desktop-client/PyStudioDesktopSharp/Services/GitService.cs
using System.Diagnostics;

namespace PyStudioDesktopSharp.Services;

public sealed record GitResult(bool Ok, string Message);

public sealed class GitService
{
    public bool IsGitAvailable() => TryRun(null, "git", "--version", 5).Ok;

    public GitResult InitRepo(string projectRoot)
    {
        if (!IsGitAvailable())
            return new GitResult(false, "Git no está instalado o no está en PATH.");

        if (Directory.Exists(Path.Combine(projectRoot, ".git")))
            return new GitResult(true, "El repositorio Git local ya existe.");

        var init = TryRun(projectRoot, "git", "init", 10);
        if (!init.Ok) return init;

        TryRun(projectRoot, "git", "config user.email pystudio.local@desktop", 5);
        TryRun(projectRoot, "git", "config user.name \"PyStudio Desktop CSharp\"", 5);
        return new GitResult(true, "Repositorio Git local inicializado.");
    }

    public GitResult CommitAll(string projectRoot, string message)
    {
        if (!IsGitAvailable())
            return new GitResult(false, "Git no está instalado o no está en PATH.");

        InitRepo(projectRoot);
        TryRun(projectRoot, "git", "add .", 10);
        var commit = TryRun(projectRoot, "git", $"commit -m \"{Escape(message)}\"", 15);

        string combined = (commit.Message ?? string.Empty).ToLowerInvariant();
        if (!commit.Ok && (combined.Contains("nothing to commit") || combined.Contains("no changes added")))
            return new GitResult(true, "No había cambios nuevos para commitear.");

        return commit.Ok
            ? new GitResult(true, $"Commit automático creado: {message}")
            : commit;
    }

    public string History(string projectRoot)
    {
        if (!IsGitAvailable())
            return "Git no está instalado o no está en PATH.";

        InitRepo(projectRoot);
        var history = TryRun(projectRoot, "git", "log --oneline --decorate --date=local --pretty=format:\"%h | %ad | %s\"", 10);
        return string.IsNullOrWhiteSpace(history.Message) ? "No hay commits todavía." : history.Message;
    }

    private static string Escape(string value) => value.Replace("\"", "'");

    private static GitResult TryRun(string? workingDirectory, string fileName, string arguments, int timeoutSeconds)
    {
        try
        {
            using var process = new Process();
            process.StartInfo = new ProcessStartInfo
            {
                FileName = fileName,
                Arguments = arguments,
                WorkingDirectory = workingDirectory ?? Environment.CurrentDirectory,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true
            };

            process.Start();
            bool exited = process.WaitForExit(timeoutSeconds * 1000);
            if (!exited)
            {
                try { process.Kill(true); } catch { }
                return new GitResult(false, "El comando Git excedió el tiempo permitido.");
            }

            string output = process.StandardOutput.ReadToEnd();
            string error = process.StandardError.ReadToEnd();
            string message = string.IsNullOrWhiteSpace(output) ? error : output + Environment.NewLine + error;
            return new GitResult(process.ExitCode == 0, message.Trim());
        }
        catch (Exception ex)
        {
            return new GitResult(false, ex.Message);
        }
    }
}
